File: src/ble/magicLanternCommands.ts
import { clampByte, clampPercent, type RgbColor } from '../color.js';
import type { BrightnessMode, PowerMode } from '../types.js';

export const MAGIC_LANTERN_DEFAULT_SERVICE_UUID = 'fff0';
export const MAGIC_LANTERN_DEFAULT_CHARACTERISTIC_UUID = 'fff3';

const frame = (...bytes: number[]): Buffer => Buffer.from(bytes.map(clampByte));

export const buildPowerCommand = (on: boolean): Buffer => {
  return on
    ? Buffer.from('7e0404f00001ff00ef', 'hex')
    : Buffer.from('7e0404000000ff00ef', 'hex');
};

export const buildColorCommand = ({ red, green, blue }: RgbColor): Buffer => {
  return frame(0x7e, 0x07, 0x05, 0x03, red, green, blue, 0x10, 0xef);
};

export const buildBrightnessCommand = (brightness: number): Buffer => {
  return frame(0x7e, 0x04, 0x01, clampPercent(brightness), 0x01, 0xff, 0xff, 0x00, 0xef);
};

export const shouldUseNativePower = (powerMode: PowerMode = 'both'): boolean => {
  return powerMode === 'native' || powerMode === 'both';
};

export const shouldUseRgbBlackOff = (powerMode: PowerMode = 'both'): boolean => {
  return powerMode === 'rgbBlack' || powerMode === 'both';
};

export const shouldUseNativeBrightness = (brightnessMode: BrightnessMode = 'rgb'): boolean => {
  return brightnessMode === 'native' || brightnessMode === 'both';
};

export const shouldScaleRgbBrightness = (brightnessMode: BrightnessMode = 'rgb'): boolean => {
  return brightnessMode === 'rgb' || brightnessMode === 'both';
};

export const buildNativeBrightnessCommands = (
  brightness: number,
  brightnessMode: BrightnessMode = 'rgb',
): Buffer[] => {
  return shouldUseNativeBrightness(brightnessMode) ? [buildBrightnessCommand(brightness)] : [];
};

export const buildPowerOnPrefixCommands = (
  powerMode: PowerMode = 'both',
  brightnessMode: BrightnessMode = 'rgb',
  brightness = 100,
): Buffer[] => {
  return [
    ...(shouldUseNativePower(powerMode) ? [buildPowerCommand(true)] : []),
    ...buildNativeBrightnessCommands(brightness, brightnessMode),
  ];
};

export const buildPowerOffCommands = (powerMode: PowerMode = 'both'): Buffer[] => {
  return [
    ...(shouldUseRgbBlackOff(powerMode) ? [buildColorCommand({ red: 0, green: 0, blue: 0 })] : []),
    ...(shouldUseNativePower(powerMode) ? [buildPowerCommand(false)] : []),
  ];
};

export const buildEffectSpeedCommand = (speed: number): Buffer => {
  return frame(0x7e, 0x04, 0x02, clampPercent(speed), 0xff, 0xff, 0xff, 0x00, 0xef);
};

export const buildBasicEffectCommand = (effectCode: number): Buffer => {
  return frame(0x7e, 0x05, 0x03, effectCode, 0x06, 0xff, 0xff, 0x00, 0xef);
};

export const buildBasicEffectCommands = (effectCode: number, speed: number): Buffer[] => {
  return [
    buildEffectSpeedCommand(speed),
    buildBasicEffectCommand(effectCode),
  ];
};

File: swift/LanternICDaemon/Sources/LanternICCore/MagicLanternCommands.swift
import Foundation

public enum MagicLanternHexError: Error, Equatable {
  case oddLength
  case invalidByte(String)
}

public enum MagicLanternCommands {
  public static let defaultServiceUUID = "fff0"
  public static let defaultCharacteristicUUID = "fff3"

  public static func power(_ isOn: Bool) -> [UInt8] {
    isOn
      ? [0x7e, 0x04, 0x04, 0xf0, 0x00, 0x01, 0xff, 0x00, 0xef]
      : [0x7e, 0x04, 0x04, 0x00, 0x00, 0x00, 0xff, 0x00, 0xef]
  }

  public static func color(red: Int, green: Int, blue: Int) -> [UInt8] {
    frame([0x7e, 0x07, 0x05, 0x03, red, green, blue, 0x10, 0xef])
  }

  public static func brightness(_ brightness: Int) -> [UInt8] {
    frame([0x7e, 0x04, 0x01, Int(clampPercent(brightness)), 0x01, 0xff, 0xff, 0x00, 0xef])
  }

  public static func effectSpeed(_ speed: Int) -> [UInt8] {
    frame([0x7e, 0x04, 0x02, Int(clampPercent(speed)), 0xff, 0xff, 0xff, 0x00, 0xef])
  }

  public static func basicEffect(_ effectCode: Int) -> [UInt8] {
    frame([0x7e, 0x05, 0x03, effectCode, 0x06, 0xff, 0xff, 0x00, 0xef])
  }

  public static func rgbBlackOff() -> [UInt8] {
    color(red: 0, green: 0, blue: 0)
  }

  public static func clampByte(_ value: Int) -> UInt8 {
    UInt8(max(0, min(255, value)))
  }

  public static func clampPercent(_ value: Int) -> UInt8 {
    UInt8(max(0, min(100, value)))
  }

  public static func hexString(for bytes: [UInt8]) -> String {
    bytes.map { String(format: "%02x", $0) }.joined()
  }

  public static func bytes(fromHex hex: String) throws -> [UInt8] {
    let cleaned = hex.trimmingCharacters(in: .whitespacesAndNewlines)
    guard cleaned.count.isMultiple(of: 2) else {
      throw MagicLanternHexError.oddLength
    }

    var bytes: [UInt8] = []
    var index = cleaned.startIndex

    while index < cleaned.endIndex {
      let nextIndex = cleaned.index(index, offsetBy: 2)
      let byteString = String(cleaned[index..<nextIndex])

      guard let byte = UInt8(byteString, radix: 16) else {
        throw MagicLanternHexError.invalidByte(byteString)
      }

      bytes.append(byte)
      index = nextIndex
    }

    return bytes
  }

  private static func frame(_ bytes: [Int]) -> [UInt8] {
    bytes.map(clampByte)
  }
}
